### PR TITLE
Added Retry Logic

### PR DESCRIPTION
**Pull request by kingforaday**
CZDS had previously shown some instability in their infrastructure, which seems to have stabilized, but should a connection timeout occur during zone download or other disconnect, this will re-add the zone location back to the processing queue for further retries.

### CODE REVIEW

File: cmd/czds-dl/download.go [hunk 1]
```diff
@@ -103,22 +110,32 @@ func main() {
 func addLinks(downloads []string) {
 	for _, dl := range downloads {
 		work.Add(1)
-		inputChan <- dl
-	}
-	close(inputChan)
```
Comment (lanrat) [inline cmd/czds-dl/download.go]: Closing `inputChan` helps prevent deadlock, was this removal intentional?

File: cmd/czds-dl/download.go [hunk 2]
```diff
@@ -103,22 +110,32 @@ func main() {
...
 				// fixes occasional HTTP 500s from CZDS
-				log.Print(err)
+				v("[%s] err: %s", path.Base(zi.Dl), err)
+				zi.Count++
+				if uint(zi.Count) < *retries {
+					work.Add(1)
+					inputChan <- zi // requeue
+				} else {
+					log.Printf("[%s] Max fail count hit; not downloading.", path.Base(zi.Dl))
+					_ = os.Remove(zi.FullPath)
```
Comment (lanrat) [inline cmd/czds-dl/download.go]: Can you add error checking here?
It does not need to be a fatal error, but printing it to stderr if something wonky is going on with the filesystem is important.

**Review by lanrat — COMMENTED** [review]
Thanks for the PR! 😃
I've made a few comments inline.

File: cmd/czds-dl/download.go (hunk 1 above)
Comment (kingforaday) [reply cmd/czds-dl/download.go]: It was intentional.  The channel is re-used later to requeue failed downloads.  Careful to add to the worker group so the task items doesn't come desynced.

Comment (lanrat) [reply cmd/czds-dl/download.go]: Unfortunately; I think that this introduced a deadlock possibility.

For example, say `client.GetLinks()` called from the `addLinks()` goroutine returns 1000 zones, and starts inserting them into `inputChan` which has a buffer of 100. Since there are more zones than fit in the buffer `addLinks()` will block when inserting into `inputChan` after it is full.
Now while `addLinks()` is blocked, if CZDS has some errors and zones need to be retried, all the threads will also block when attempting to insert the zones back into `inputChan` because `inputChan` is still at capacity, resulting in deadlock.

I'm trying to think of an elegant way to fix this. One possibility is to run the line `inputChan <- zi // requeue` in its own goroutine to not block the worker thread, but this may have the unintended consequence of starting a separate thread for every zone if CZDS is having a bad day.

Comment (kingforaday) [reply cmd/czds-dl/download.go]: I think I understand your point.  You are saying that if and only if addLinks still has entries to add, but also all workers have failed on their fetch and try to re-add to the queue would you experience a deadlock.  Realistically, I can only imagine this scenario getting hit when CZDS's backend is having a major hiccup.  Let me give this some thought as well.

**Comment by kingforaday** [comment]
Take a look at the latest commit [cb6a11a](https://github.com/lanrat/czds/pull/3/commits/cb6a11ab61ab01a65e704b7a751f0e6dbe8c28bf).  To not interrupt the addLinks channel management there is a separate one for any retries required.

**Comment by lanrat** [comment]
Unfortunately this does not fix the deadlock either.

I tested your branch returning an error randomly on calls to `zoneDownload(zi)` and after the first few zones the program stopped in a deadlock.

The problem is that once the `retryChan` is full we are back where we started where the worker threads are blocked adding new items to the channel and cant take items out of either `retryChan` or `inputChan`.

**Comment by lanrat** [comment]
The best solution I can think of is to take the code from commit a7ea2867fb9cc34ed83c8c84a1688e92d602bf94 and change the line: 
```
inputChan <- zi // requeue
```

to:
```
go func() {
    inputChan <- zi // requeue
}()
```
This adds the `ZoneInfo` back to  `inputChan` in its own goroutine and does not block the worker. Its not ideal, but it will never deadlock.

**Comment by lanrat** [comment]
This has been added in #4
